File: src/lib/workflows/contracts.ts
// src/lib/workflows/contracts.ts
"use server";

import { callGemini, streamGemini, cleanJsonString } from "../utils";

/**
 * Extract contract clauses from a chunk of text
 * @param chunk Text chunk to analyze
 * @returns Generator yielding extracted clauses
 */
async function* extractContractClauses(chunk: string) {
  const prompt = `
    You are a legal expert specializing in contract analysis. Extract all contract clauses from this text.
    Return each clause on a new line. Focus on identifying distinct contractual provisions, obligations, rights, or terms.
    
    Text to analyze:
    ${chunk}
  `;
  
  yield* streamGemini(prompt);
}

/**
 * Classify a contract clause into a subcategory
 * @param clause The contract clause to classify
 * @returns Generator yielding the classification
 */
async function* classifyContractLevel2(clause: string) {
  const prompt = `
    You are a legal expert specializing in contract analysis. Classify this contract clause into ONE of these categories:
    
    1. Payment Terms
    2. Termination
    3. Confidentiality
    4. Intellectual Property
    5. Liability
    6. Warranties
    7. Force Majeure
    8. Dispute Resolution
    9. Assignment
    10. Governing Law
    11. Other
    
    Return ONLY the category name, nothing else.
    
    Clause to classify:
    ${clause}
  `;
  
  yield* streamGemini(prompt);
}

/**
 * Extract attributes from a contract clause
 * @param clause The contract clause to analyze
 * @returns Generator yielding extracted attributes
 */
async function* extractContractAttributes(clause: string) {
  const prompt = `
    You are a legal expert specializing in contract analysis. Extract key attributes from this contract clause.
    
    For each clause, provide:
    1. The parties involved (if mentioned)
    2. Any specific obligations or rights
    3. Any conditions or triggers
    4. Any deadlines or time periods
    5. Any monetary values or percentages
    
    Return the results in JSON format with these fields:
    {
      "parties": [list of parties involved],
      "obligations": [list of obligations],
      "conditions": [list of conditions],
      "timeframes": [list of deadlines or periods],
      "values": [list of monetary values or percentages]
    }
    
    Clause to analyze:
    ${clause}
  `;
  
  yield* streamGemini(prompt);
}

/**
 * Explain a contract clause in plain language
 * @param clause The contract clause to explain
 * @returns Generator yielding the explanation
 */
async function* explainContractClause(clause: string) {
  const prompt = `
    You are a legal expert specializing in making contracts understandable to non-lawyers.
    Explain this contract clause in plain, simple language that a high school student could understand.
    
    Focus on:
    1. What this clause means in practical terms
    2. What obligations it creates
    3. What risks or benefits it presents
    
    Clause to explain:
    ${clause}
  `;
  
  yield* streamGemini(prompt);
}

/**
 * Generate a summary of contract analysis
 * @param contractResults The contract analysis results
 * @returns Generator yielding the summary
 */
async function* generateContractSummary(contractResults: any[]) {
  const prompt = `
    You are a legal expert specializing in contract analysis. Generate a comprehensive summary of this contract based on the analyzed clauses.
    
    Focus on:
    1. The overall purpose and structure of the contract
    2. Key rights and obligations of each party
    3. Important deadlines or milestones
    4. Potential risks or areas of concern
    5. Recommendations for the reader
    
    Analyzed clauses:
    ${JSON.stringify(contractResults, null, 2)}
    
    Provide a well-structured summary with clear headings and bullet points where appropriate.
  `;
  
  yield* streamGemini(prompt);
}

/**
 * Process a contract document
 * @param chunks Text chunks from the document
 * @returns Promise resolving to [summary, jsonData]
 */
export async function contractWorkflow(chunks: string[]): Promise<[string, any]> {
  console.log("Starting contract workflow...");
  
  // Extract all clauses from chunks
  const allClauses: string[] = [];
  
  console.log("Extracting contract clauses in streaming mode...\n");
  for (const chunk of chunks) {
    let clauseText = "";
    for await (const partial of extractContractClauses(chunk)) {
      process.stdout.write(partial); // Stream to terminal
      clauseText += partial;
    }
    
    const clauses = clauseText.split("\n").filter(c => c.trim());
    allClauses.push(...clauses);
  }
  
  // Deduplicate clauses
  const uniqueClauses = Array.from(new Set(allClauses));
  
  // Classify clauses
  const contractResults = [];
  for (const clause of uniqueClauses) {
    let subCategory = "";
    for await (const part of classifyContractLevel2(clause)) {
      process.stdout.write(part); // Stream to terminal
      subCategory += part;
    }
    
    contractResults.push({
      clause,
      subCategory: subCategory.trim()
    });
  }
  
  // Extract attributes
  const contractAttributes = [];
  for (const clause of uniqueClauses) {
    let details = "";
    for await (const part of extractContractAttributes(clause)) {
      process.stdout.write(part); // Stream to terminal
      details += part;
    }
    
    contractAttributes.push({
      clause,
      attributes: cleanJsonString(details)
    });
  }
  
  // Explain clauses
  const explainedClauses = [];
  for (const clause of uniqueClauses) {
    let explanation = "";
    for await (const part of explainContractClause(clause)) {
      process.stdout.write(part); // Stream to terminal
      explanation += part;
    }
    
    explainedClauses.push({
      clause,
      explanation: explanation.trim()
    });
  }
  
  // Generate summary
  let summary = "";
  for await (const part of generateContractSummary([...contractResults, ...contractAttributes, ...explainedClauses])) {
    process.stdout.write(part); // Stream to terminal
    summary += part;
  }
  
  // Combine all data for JSON response
  const jsonData = {
    clauses: uniqueClauses.length,
    classifications: contractResults,
    attributes: contractAttributes,
    explanations: explainedClauses
  };
  
  return [summary, jsonData];
}

File: contracts.py
from utils import *
import json
import re

class contracts:

    def __init__(self,chunks):
        self.chunks = chunks


    def extract_contract_clauses(self, chunk_text):
        prompt = f"""
        You are a legal contract analyzer. 
        Your task is to read the given contract/agreement text and extract **both Parties and Clauses**.

        Rules:
        1. First, list the **Parties** involved in the contract with their roles (e.g., Employer – ABC Corp, Employee – John Doe, Landlord – Mr. X, Tenant – Ms. Y).
        2. After that, list the **major Clauses**. A clause means any distinct contractual provision such as "Confidentiality", "Termination", "Governing Law", "Payment Terms", etc.
        3. Each clause should be summarized into **one concise sentence**.
        4. Format strictly as a **clean numbered list** (1., 2., 3., …).  
        5. Do not explain, do not repeat, do not add anything else. Just return the numbered list.

        Document Text:
        {chunk_text}
        """
        return call_gemini1(prompt)


    def classify_contract_level2(self,clause_text):
        """Classify each clause into Contract sub-categories"""
        
        prompt = f"""
        You are a legal assistant specializing in Contract Law.

        Classify the following clause into one of these sub-categories 
        (or suggest a new but precise sub-category if none fits):


        - Core Relationship: Parties, Scope, Duration
        - Financial Terms: Salary, Fees, Rent, Payment Terms
        - Performance & Obligations: Duties, Service Levels, Maintenance
        - Confidentiality & IP: NDA, Trade Secrets, IP Ownership
        - Termination & Exit: Termination grounds, Notice periods, Handover
        - Risk & Restrictions: Indemnity, Liability, Non-Compete, Non-Solicitation
        - Dispute Handling: Governing Law, Arbitration, Mediation, Jurisdiction
        - Boilerplate: Force Majeure, Severability, Entire Agreement, Notices

        Clause:
        \"\"\"{clause_text}\"\"\"  

        Return only the sub-category name.
        """
        return call_gemini1(prompt)




    def extract_contract_attributes(self, clause_text):
        """Extract structured attributes for Contract clauses"""

        prompt = f"""
        You are a legal assistant extracting structured data from contract clauses.

        From the following clause, extract the attributes in **JSON format** with these keys:
        - Parties
        - Scope
        - FinancialTerms
        - Obligations
        - Confidentiality
        - IP_Rights
        - TerminationConditions
        - RiskRestrictions
        - DisputeResolution
        - Boilerplate
        - OtherNotes

        If an attribute is not relevant, return null.

        Clause:
        \"\"\"{clause_text}\"\"\"
        """
        return call_gemini1(prompt)




    def explain_contract_clause(self, clause_text):
        """Generate explanation + practical effect for a Contract clause"""

        prompt = f"""
        You are a legal assistant.

        TASK: Read the clause and provide:
        1. A clear explanation of what the clause states in plain English.
        2. The practical effect on the parties (rights, obligations, benefits, restrictions).

        Rules:
        - Focus only on the given clause, do not assume or speculate about missing terms.
        - Cover everything explicitly mentioned in the clause.
        - Keep language clear, professional, and non-repetitive.
        - Explanations can be a few sentences if needed, but avoid unnecessary length.
        - If the clause only identifies a party or role, state that directly without judging completeness.

        Clause:
        \"\"\"{clause_text}\"\"\"

        Return the output in JSON format with every key in diffrent line:
        - Explanation
        - PracticalEffect
        """
        return call_gemini1(prompt)
    

    def makenice(self, clause_text):
        """Generate explanation + practical effect for a Contract clause"""

        prompt = f"""
        We would give you a json file and make it display nice and which cell are duplicated make it as one and remove all null columns and the key name which are repeated merge them in a list display it nicely 
        {clause_text}
        """
        return call_gemini1(prompt)

    